File: services/runtime/agent.py
"""Isolate agent.

One of these runs inside every isolate. It imports the function once, then
answers invocation requests from the control plane over the function network.
Standard library only — the less this has to import, the shorter the cold start.

Protocol
    GET  /healthz  -> {"ready": bool, "error": str|null, "fatal": bool}
    POST /invoke   -> {"status_code", "body", "headers", "logs",
                       "context_writes", "context_deletes", "error"}
"""

from __future__ import annotations

import importlib.util
import inspect
import io
import json
import os
import sys
import threading
import time
import traceback
from concurrent.futures import ThreadPoolExecutor, TimeoutError as FutureTimeout
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

import cubicle_context
from cubicle_context import Context, Request

PORT = int(os.environ.get("CUBICLE_AGENT_PORT", "8080"))
SOURCE = os.environ.get("CUBICLE_HANDLER_PATH", "/srv/handler.py")
ENTRYPOINT = os.environ.get("CUBICLE_ENTRYPOINT", "handler")
MAX_LOG_LINES = 200

_handler = None
_load_error: str | None = None
_executor = ThreadPoolExecutor(max_workers=8, thread_name_prefix="invoke")


# ── per-invocation stdout capture ────────────────────────────────────────────


class _StreamRouter(io.TextIOBase):
    """Routes writes to the current invocation's buffer, else to the real stream."""

    def __init__(self, real, level: str) -> None:
        self._real = real
        self._level = level
        self._local = threading.local()

    def begin(self) -> list[dict]:
        buffer: list[dict] = []
        self._local.buffer = buffer
        return buffer

    def end(self) -> None:
        self._local.buffer = None

    def write(self, text: str) -> int:
        buffer = getattr(self._local, "buffer", None)
        if buffer is None:
            return self._real.write(text)
        for line in text.splitlines():
            if line and len(buffer) < MAX_LOG_LINES:
                buffer.append({"level": self._level, "message": line[:4000]})
        return len(text)

    def flush(self) -> None:
        self._real.flush()


_stdout = _StreamRouter(sys.stdout, "INFO")
_stderr = _StreamRouter(sys.stderr, "ERROR")
sys.stdout = _stdout
sys.stderr = _stderr


# ── function loading ─────────────────────────────────────────────────────────


def load_function() -> None:
    global _handler, _load_error
    try:
        if not os.path.exists(SOURCE):
            raise FileNotFoundError(f"{SOURCE} is missing from the deployed bundle")

        spec = importlib.util.spec_from_file_location("cubicle_function", SOURCE)
        if spec is None or spec.loader is None:
            raise ImportError(f"could not load {SOURCE}")
        module = importlib.util.module_from_spec(spec)
        sys.modules["cubicle_function"] = module
        spec.loader.exec_module(module)

        fn = getattr(module, ENTRYPOINT, None)
        if fn is None or not callable(fn):
            raise AttributeError(f"{SOURCE} defines no callable named '{ENTRYPOINT}'")
        _handler = fn
        _load_error = None
    except Exception:
        _load_error = traceback.format_exc(limit=6)
        _handler = None


# ── invocation ───────────────────────────────────────────────────────────────


def normalise(result) -> tuple[int, object, dict]:
    """Accept every documented return shape and produce one response."""
    if isinstance(result, tuple) and len(result) == 2:
        body, status = result
        return int(status), body, {}
    if isinstance(result, dict):
        status = result.get("statusCode", result.get("status_code"))
        if status is not None:
            headers = result.get("headers") or {}
            body = result.get("body")
            if isinstance(body, str):
                try:
                    body = json.loads(body)
                except (TypeError, ValueError):
                    pass
            return int(status), body, {str(k): str(v) for k, v in headers.items()}
    return 200, result, {}


def run_invocation(payload: dict) -> dict:
    if _handler is None:
        return {
            "status_code": 500,
            "body": {"error": "handler_not_loaded", "message": _load_error or "unknown"},
            "headers": {},
            "logs": [{"level": "ERROR", "message": (_load_error or "").strip()[:4000]}],
            "error": "handler failed to import",
            "context_writes": {},
            "context_deletes": [],
        }

    request = Request(payload)
    context = Context(
        payload.get("context") or {},
        payload.get("ctx_access", "rw"),
        payload.get("session_id", ""),
    )
    cubicle_context._bind(
        env_values=payload.get("env") or {},
        secrets=payload.get("secrets") or {},
        services=payload.get("services") or {},
    )

    logs = _stdout.begin()
    _stderr._local.buffer = logs
    started = time.perf_counter()
    error = None
    try:
        parameters = len(inspect.signature(_handler).parameters)
        result = _handler(request, context) if parameters >= 2 else _handler(request)
        status_code, body, headers = normalise(result)
    except Exception as exc:
        trace = traceback.format_exc(limit=8)
        error = f"{exc.__class__.__name__}: {exc}"
        status_code, body, headers = 500, {"error": "handler_error", "message": str(exc)}, {}
        for line in trace.strip().splitlines():
            if len(logs) < MAX_LOG_LINES:
                logs.append({"level": "ERROR", "message": line[:4000]})
    finally:
        _stdout.end()
        _stderr.end()
        cubicle_context._clear()

    writes, deletes = context._drain()
    return {
        "status_code": status_code,
        "body": body,
        "headers": headers,
        "logs": logs,
        "error": error,
        "duration_ms": round((time.perf_counter() - started) * 1000, 2),
        "context_writes": writes,
        "context_deletes": deletes,
    }


# ── HTTP ─────────────────────────────────────────────────────────────────────


class Agent(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"
    server_version = "cubicle-agent/1.0"

    def _send(self, status: int, payload: dict) -> None:
        body = json.dumps(payload, default=str).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self) -> None:  # noqa: N802 - BaseHTTPRequestHandler API
        if self.path.startswith("/healthz"):
            self._send(
                200,
                {
                    "ready": _handler is not None,
                    "error": _load_error,
                    "fatal": _handler is None and _load_error is not None,
                    "function": os.environ.get("CUBICLE_FUNCTION", ""),
                },
            )
        else:
            self._send(404, {"error": "not_found"})

    def do_POST(self) -> None:  # noqa: N802 - BaseHTTPRequestHandler API
        if not self.path.startswith("/invoke"):
            self._send(404, {"error": "not_found"})
            return

        length = int(self.headers.get("Content-Length", "0") or 0)
        try:
            payload = json.loads(self.rfile.read(length) or b"{}")
        except ValueError:
            self._send(400, {"error": "bad_request", "message": "payload is not valid JSON"})
            return

        timeout = float(payload.get("timeout_s") or os.environ.get("CUBICLE_TIMEOUT", "30"))
        future = _executor.submit(run_invocation, payload)
        try:
            self._send(200, future.result(timeout=timeout))
        except FutureTimeout:
            # The handler is still running. The control plane recycles this
            # isolate on a 504, which is the only way to reclaim the thread.
            self._send(
                200,
                {
                    "status_code": 504,
                    "body": {
                        "error": "timeout",
                        "message": f"handler exceeded its {timeout:g}s timeout",
                    },
                    "headers": {},
                    "logs": [{"level": "ERROR", "message": f"timed out after {timeout:g}s"}],
                    "error": "timeout",
                    "context_writes": {},
                    "context_deletes": [],
                },
            )

    def log_message(self, *args) -> None:  # noqa: A003 - silence the default access log
        return


def main() -> None:
    load_function()
    if _load_error:
        # Reported through /healthz; the control plane surfaces it in the console.
        print(f"cubicle-agent: function failed to load\n{_load_error}", file=sys.__stderr__)
    server = ThreadingHTTPServer(("0.0.0.0", PORT), Agent)
    server.daemon_threads = True
    server.serve_forever()


if __name__ == "__main__":
    main()

File: services/runtime/node/agent.js
/**
 * Isolate agent — JavaScript.
 *
 * The Node counterpart of agent.py, speaking the identical protocol so the
 * control plane cannot tell the two apart:
 *
 *   GET  /healthz  -> {"ready": bool, "error": string|null, "fatal": bool}
 *   POST /invoke   -> {"status_code", "body", "headers", "logs",
 *                      "context_writes", "context_deletes", "error"}
 *
 * Node built-ins only. The less this has to load, the shorter the cold start —
 * the same reasoning that keeps the Python agent on the standard library.
 */

'use strict'

const http = require('node:http')
const path = require('node:path')
const { pathToFileURL } = require('node:url')

const PORT = Number(process.env.CUBICLE_AGENT_PORT || 8080)
const SOURCE = process.env.CUBICLE_HANDLER_PATH || '/srv/handler.js'
const ENTRYPOINT = process.env.CUBICLE_ENTRYPOINT || 'handler'
const MAX_LOG_LINES = 200

let handler = null
let loadError = null

// ── loading ──────────────────────────────────────────────────────────────────

/**
 * Resolve the exported handler across the shapes people actually write:
 * `export function handler`, `export default`, and `module.exports =`.
 */
function pick(mod) {
  if (!mod) return null
  if (typeof mod[ENTRYPOINT] === 'function') return mod[ENTRYPOINT]
  const fallback = mod.default
  if (typeof fallback === 'function') return fallback
  if (fallback && typeof fallback[ENTRYPOINT] === 'function') return fallback[ENTRYPOINT]
  return null
}

async function load() {
  try {
    // A dynamic import handles ESM and CommonJS alike; require() would refuse
    // the first, which is what most people write now.
    const mod = await import(pathToFileURL(path.resolve(SOURCE)).href)
    handler = pick(mod)
    if (!handler) {
      loadError =
        `${SOURCE} does not export '${ENTRYPOINT}'. ` +
        `Export it by name, or as the default export.`
    }
  } catch (err) {
    loadError = (err && err.stack) || String(err)
    handler = null
  }
}

// ── per-invocation console capture ───────────────────────────────────────────
//
// A module-level buffer is safe here: the pool marks an isolate busy for the
// duration of a request and only ever hands out idle ones, so exactly one
// invocation runs in this process at a time.

let logs = null
const real = {
  log: console.log,
  info: console.info,
  warn: console.warn,
  error: console.error,
  debug: console.debug,
}

function record(level, args) {
  const message = args
    .map((value) => {
      if (typeof value === 'string') return value
      try {
        return JSON.stringify(value)
      } catch {
        return String(value)
      }
    })
    .join(' ')

  for (const line of message.split('\n')) {
    if (logs && logs.length < MAX_LOG_LINES) logs.push({ level, message: line.slice(0, 4000) })
  }
}

function captureConsole() {
  console.log = (...args) => (logs ? record('INFO', args) : real.log(...args))
  console.info = (...args) => (logs ? record('INFO', args) : real.info(...args))
  console.debug = (...args) => (logs ? record('DEBUG', args) : real.debug(...args))
  console.warn = (...args) => (logs ? record('WARN', args) : real.warn(...args))
  console.error = (...args) => (logs ? record('ERROR', args) : real.error(...args))
}

// ── request and context ──────────────────────────────────────────────────────

class Request {
  constructor(payload) {
    this.body = payload.body ?? null
    this.headers = { ...(payload.headers || {}) }
    this.method = payload.method || 'POST'
    this.query = { ...(payload.query || {}) }
    this.path = payload.path || '/'
    this.sessionId = payload.session_id || ''
    this.requestId = payload.request_id || ''
    this.namespace = payload.namespace || ''
    this.function = payload.function || ''
    this.trigger = payload.trigger || 'http'
  }

  /** The body as an object, whether it arrived parsed or as a JSON string. */
  json() {
    if (this.body === null || this.body === '') return null
    if (typeof this.body === 'object') return this.body
    try {
      return JSON.parse(this.body)
    } catch {
      return this.body
    }
  }

  text() {
    if (this.body === null) return ''
    return typeof this.body === 'string' ? this.body : JSON.stringify(this.body)
  }

  header(name, fallback = null) {
    return this.headers[String(name).toLowerCase()] ?? fallback
  }
}

class ContextAccessError extends Error {}

class Context {
  constructor(data, access, sessionId) {
    this._data = { ...(data || {}) }
    this._writes = {}
    this._deletes = []
    this._access = access || 'rw'
    this.sessionId = sessionId || ''
  }

  get readable() {
    return this._access.includes('r')
  }

  get writable() {
    return this._access.includes('w')
  }

  get(key, fallback = null) {
    if (!this.readable) {
      throw new ContextAccessError(
        `this function has '${this._access}' context access and cannot read the context`,
      )
    }
    return key in this._data ? this._data[key] : fallback
  }

  set(key, value) {
    if (!this.writable) {
      throw new ContextAccessError(
        `this function has '${this._access}' context access and cannot write the context`,
      )
    }
    this._data[key] = value
    this._writes[key] = value
    const at = this._deletes.indexOf(key)
    if (at !== -1) this._deletes.splice(at, 1)
  }

  delete(key) {
    if (!this.writable) {
      throw new ContextAccessError(
        `this function has '${this._access}' context access and cannot write the context`,
      )
    }
    delete this._data[key]
    delete this._writes[key]
    if (!this._deletes.includes(key)) this._deletes.push(key)
  }

  keys() {
    if (!this.readable) return []
    return Object.keys(this._data)
  }

  all() {
    if (!this.readable) {
      throw new ContextAccessError(
        `this function has '${this._access}' context access and cannot read the context`,
      )
    }
    return { ...this._data }
  }

  _drain() {
    const writes = this._writes
    const deletes = this._deletes
    this._writes = {}
    this._deletes = []
    return [writes, deletes]
  }
}

// ── invocation ───────────────────────────────────────────────────────────────

/** Accept every documented return shape and produce one response. */
function normalise(result) {
  if (result && typeof result === 'object' && !Array.isArray(result)) {
    const status = result.statusCode ?? result.status_code
    if (status !== undefined && status !== null) {
      let body = result.body
      if (typeof body === 'string') {
        try {
          body = JSON.parse(body)
        } catch {
          /* a plain string body is a legitimate answer */
        }
      }
      const headers = {}
      for (const [key, value] of Object.entries(result.headers || {})) {
        headers[String(key)] = String(value)
      }
      return [Number(status), body ?? null, headers]
    }
  }
  return [200, result ?? null, {}]
}

async function runInvocation(payload) {
  if (!handler) {
    return {
      status_code: 500,
      body: { error: 'handler_not_loaded', message: loadError || 'unknown' },
      headers: {},
      logs: [{ level: 'ERROR', message: String(loadError || '').slice(0, 4000) }],
      error: 'handler failed to import',
      context_writes: {},
      context_deletes: [],
    }
  }

  const request = new Request(payload)
  const context = new Context(payload.context || {}, payload.ctx_access || 'rw', payload.session_id)

  // The same three collections the Python SDK exposes, as plain frozen objects
  // rather than a module import — a handler reads them off the context.
  context.env = Object.freeze({ ...(payload.env || {}) })
  context.secrets = Object.freeze({ ...(payload.secrets || {}) })
  context.services = Object.freeze({ ...(payload.services || {}) })

  logs = []
  const started = process.hrtime.bigint()
  let error = null
  let statusCode = 200
  let body = null
  let headers = {}

  try {
    const result = await handler(request, context)
    ;[statusCode, body, headers] = normalise(result)
  } catch (err) {
    error = `${(err && err.name) || 'Error'}: ${(err && err.message) || String(err)}`
    statusCode = 500
    body = { error: 'handler_error', message: (err && err.message) || String(err) }
    headers = {}
    for (const line of String((err && err.stack) || err).split('\n')) {
      if (logs.length < MAX_LOG_LINES) logs.push({ level: 'ERROR', message: line.slice(0, 4000) })
    }
  }

  const [writes, deletes] = context._drain()
  const collected = logs
  logs = null

  return {
    status_code: statusCode,
    body,
    headers,
    logs: collected,
    error,
    duration_ms: Number(process.hrtime.bigint() - started) / 1e6,
    context_writes: writes,
    context_deletes: deletes,
  }
}

// ── HTTP ─────────────────────────────────────────────────────────────────────

function send(res, status, payload) {
  const data = Buffer.from(JSON.stringify(payload))
  res.writeHead(status, { 'Content-Type': 'application/json', 'Content-Length': data.length })
  res.end(data)
}

const server = http.createServer((req, res) => {
  if (req.method === 'GET' && req.url.startsWith('/healthz')) {
    // `fatal` tells the control plane not to keep waiting: the module threw on
    // import and no amount of patience will change that.
    send(res, handler ? 200 : 503, {
      ready: Boolean(handler),
      error: loadError,
      fatal: Boolean(loadError),
    })
    return
  }

  if (req.method === 'POST' && req.url.startsWith('/invoke')) {
    const chunks = []
    req.on('data', (chunk) => chunks.push(chunk))
    req.on('end', async () => {
      let payload
      try {
        payload = JSON.parse(Buffer.concat(chunks).toString() || '{}')
      } catch (err) {
        send(res, 400, { error: 'bad_request', message: String(err) })
        return
      }
      try {
        send(res, 200, await runInvocation(payload))
      } catch (err) {
        // The agent itself failed, not the handler. Say so plainly rather than
        // leaving the control plane on a socket that never answers.
        send(res, 500, {
          status_code: 500,
          body: { error: 'agent_error', message: String((err && err.message) || err) },
          headers: {},
          logs: [],
          error: String(err),
          context_writes: {},
          context_deletes: [],
        })
      }
    })
    return
  }

  send(res, 404, { error: 'not_found' })
})

captureConsole()
load().then(() => {
  server.listen(PORT, '0.0.0.0', () => real.log(`cubicle agent listening on ${PORT}`))
})

// Keep serving: a rejected promise somewhere in user code must not take the
// isolate down mid-request and turn one failed call into a cold start.
process.on('unhandledRejection', (reason) => real.error('unhandled rejection:', reason))
